src/Compiler/Driver/CompilerImports.fs

```diff
@@ -2233,6 +2253,9 @@ and [<Sealed>] TcImports
             let _dllinfos, phase2s = results |> Array.choose id |> List.ofArray |> List.unzip
             fixupOrphanCcus ()
             let ccuinfos = List.collect (fun phase2 -> phase2 ()) phase2s
+            if importsBase.IsSome then
+                importsBase.Value.CcuTable.Values |> Seq.iter addConstraintSources
+                ccuTable.Values |> Seq.iter addConstraintSources
```

Comment:
That's the core of the fix.
We want the source of the constraint (the operator/member/function that defines it) in the error message. At the place where the error message is created, there is today no way to find the source. The only way (I see) to provide the sources is to add them beforehand to the constraints by going through all imports. (I also found it cannot be done during import due to the way the metadata are structured, so I had to do it after import.)
Yes, this is quite an effort for some improvement of an error message. I would think it is worth it, but that is not for me to decide in the end. And it certainly depends on the performance implications.
Regarding performance implications, I think we have to check it. Since I have never done that, can you give me a hint on how to do that or a PR that has done something similar?

Reply:
I found the benchmark section in the DevGuide. Let me try that.

Reply:
I'm more interested in high level, like which paths include new ccu rebuilding? Is it once on-load, or every time we check project?

Reply:
Just once on load. Note also that it is not rebuilding the CCUs, but just walking them to set the new mutable "source" field in the TraitConstraintInfo of the typars.

But nevertheless, I think I found my next task :-). I will try to find my way to create a benchmark for this fix, just out of interest. This might then enable me to help with #16407.